lib/galaxy/tools/actions/__init__.py

```diff
@@ -903,13 +903,34 @@ def _wrapped_params(
         wrapped_params = WrappedParameters(trans, tool, incoming, input_datasets=input_datasets)
         return wrapped_params
 
-    def _get_on_text(self, inp_data):
+    def _get_on_text(self, inp_data, inp_dataset_collections):
```

Comment:
I think it would help a lot to add type hints to input_data and inp_dataset_collections

Reply:
Indeed. Would it help to type them as `LegacyUnprefixedDict`? Or would something like `LegacyUnprefixedDict[str, HDA]` just work?

Reply:
`LegacyUnprefixedDict` is a detail that I don't think should matter. I'd use MutableMapping["str", TheRealTypes], where `TheRealTypes` is going to be the challenge. Probably `Union[HistoryDatasetAssociation, LibraryDatasetDatasetAssociation, HistoryDatasetCollectionAssociation, DatasetCollection and DatasetCollectionElement]` ?

Reply:
I would prefer to have this in another PR. Seems to big to be integrated here.